zkevm-circuits/src/evm_circuit/execution/precompiles/identity.rs

```diff
@@ -0,0 +1,257 @@
+use bus_mapping::circuit_input_builder::Call;
+use eth_types::{evm_types::GasCost, Field, ToScalar};
+use gadgets::util::{select, Expr};
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::{N_BYTES_MEMORY_WORD_SIZE, N_BYTES_WORD},
+        step::ExecutionState,
+        util::{
+            common_gadget::RestoreContextGadget, constraint_builder::EVMConstraintBuilder,
+            math_gadget::ConstantDivisionGadget, CachedRegion, Cell,
+        },
+    },
+    table::CallContextFieldTag,
+    witness::{Block, ExecStep, Transaction},
+};
+
+#[derive(Clone, Debug)]
+pub struct IdentityGadget<F> {
+    input_word_size: ConstantDivisionGadget<F, N_BYTES_MEMORY_WORD_SIZE>,
+    is_success: Cell<F>,
+    callee_address: Cell<F>,
+    caller_id: Cell<F>,
+    call_data_offset: Cell<F>,
+    call_data_length: Cell<F>,
+    return_data_offset: Cell<F>,
+    return_data_length: Cell<F>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for IdentityGadget<F> {
+    const EXECUTION_STATE: ExecutionState = ExecutionState::PrecompileIdentity;
+
+    const NAME: &'static str = "IDENTITY";
+
+    fn configure(cb: &mut EVMConstraintBuilder<F>) -> Self {
+        let [is_success, callee_address, caller_id, call_data_offset, call_data_length, return_data_offset, return_data_length] =
+            [
+                CallContextFieldTag::IsSuccess,
+                CallContextFieldTag::CalleeAddress,
+                CallContextFieldTag::CallerId,
+                CallContextFieldTag::CallDataOffset,
+                CallContextFieldTag::CallDataLength,
+                CallContextFieldTag::ReturnDataOffset,
+                CallContextFieldTag::ReturnDataLength,
+            ]
+            .map(|tag| cb.call_context(None, tag));
+
+        let input_word_size = ConstantDivisionGadget::construct(
+            cb,
+            call_data_length.expr() + (N_BYTES_WORD - 1).expr(),
+            N_BYTES_WORD as u64,
+        );
+
+        let gas_cost = select::expr(
+            is_success.expr(),
+            GasCost::PRECOMPILE_IDENTITY_BASE.expr()
+                + input_word_size.quotient() * GasCost::PRECOMPILE_IDENTITY_PER_WORD.expr(),
+            cb.curr.state.gas_left.expr(),
+        );
+
+        cb.precompile_info_lookup(
+            cb.execution_state().as_u64().expr(),
+            callee_address.expr(),
+            cb.execution_state().precompile_base_gas_cost().expr(),
+        );
+
+        let restore_context = RestoreContextGadget::construct2(
+            cb,
+            is_success.expr(),
+            gas_cost.expr(),
+            0.expr(),
+            0x00.expr(),             // ReturnDataOffset
+            call_data_length.expr(), // ReturnDataLength
```

Comment:
Generally the return data should depend on whether or not the call succeeded, i.e.
```rust
select::expr(is_success.expr(), call_data_length.expr(), 0x00.expr()),
```

In the case of `Identity` precompile, the only failure is in the case of insufficient gas for the call, which is diverted and handled in the `ErrorOogPrecompile` gadget.

So adding a `select` here would be less efficient, as we expect execution that's verified under this specific gadget to succeed. However, this deserves an explanation as a comment.

Reply:
Notes added.